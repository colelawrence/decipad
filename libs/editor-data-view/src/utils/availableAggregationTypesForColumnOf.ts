import { Result } from '@decipad/computer';
import { AggregationKind } from '../types';

export const availableAggregationTypesForColumnOf = (
  columnType: Result.Result['type']
): AggregationKind[] => {
  switch (columnType.kind) {
    case 'boolean':
      return ['frequency'];
    case 'date':
    case 'number':
      return ['sum', 'average', 'min', 'max', 'median', 'span', 'frequency'];
    case 'string':
      return ['frequency'];
    default:
      return [];
  }
};
