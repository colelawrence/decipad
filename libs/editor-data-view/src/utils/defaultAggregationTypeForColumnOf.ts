import { Result } from '@decipad/computer';
import { AggregationKind } from '../types';

export const defaultAggregationTypeForColumnOf = (
  columnType: Result.Result['type']
): AggregationKind | undefined => {
  switch (columnType.kind) {
    case 'boolean':
      return 'frequency';
    case 'date':
    case 'number':
      return 'sum';
    case 'string':
      return 'frequency';
    default:
      return undefined;
  }
};
