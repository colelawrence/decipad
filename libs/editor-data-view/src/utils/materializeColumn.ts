import { all, slice } from '@decipad/generator-utils';
import { type Result } from '@decipad/language-interfaces';

export type AnyImmaterializedColumn = {
  value: Result.ResultColumn;
};

export type AnyMaterializedColumn = {
  value: Result.OneResult[];
};

export type AnyColumn = AnyImmaterializedColumn | AnyMaterializedColumn;

export type ResultingMaterializedColumn<T> = Omit<T, 'value'> &
  AnyMaterializedColumn;

const MAX_COLUMN_LENGTH = 100_000;

export const materializeColumn = async <
  T extends AnyColumn,
  R = ResultingMaterializedColumn<T>
>(
  column: T
): Promise<R> => {
  if (Array.isArray(column.value)) {
    return column as unknown as R;
  }
  const value = await all(slice(column.value(), 0, MAX_COLUMN_LENGTH + 1));
  if (value.length > MAX_COLUMN_LENGTH) {
    throw new Error(
      `Maximum column length of ${MAX_COLUMN_LENGTH} reached. Bailing.`
    );
  }
  return {
    ...column,
    value,
  } as R;
};
