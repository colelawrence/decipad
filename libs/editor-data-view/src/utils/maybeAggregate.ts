import { columnAggregationTypes } from '@decipad/editor-table';
import { SerializedType } from '@decipad/computer';
import { TableCellType } from '@decipad/editor-types';
import { AggregationKind } from '../types';

export const maybeAggregate = (
  expressionFilter: string,
  columnType: SerializedType,
  aggregation: AggregationKind | undefined
): string | Error | undefined => {
  try {
    if (!aggregation) {
      return undefined;
    }
    const aggregations = columnAggregationTypes(columnType as TableCellType);
    const aggregator = aggregations.find((agg) => agg.name === aggregation);
    if (!aggregator) {
      throw new Error(
        `No aggregator named ${aggregation} for column type ${columnType.kind} found`
      );
    }

    return aggregator.expression(expressionFilter);
  } catch (err) {
    return err as Error;
  }
};
