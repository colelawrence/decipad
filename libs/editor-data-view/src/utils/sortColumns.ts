import { Column as ColumnImpl, Comparable, applyMap } from '@decipad/column';
import { all, filter } from '@decipad/generator-utils';
import { Column } from '../types';

export const sortColumns = async (
  availableColumns: Column[] | undefined,
  columnMap?: number[]
): Promise<Column[] | undefined> => {
  if (!columnMap || !availableColumns) {
    return undefined;
  }

  const sortedColumns = applyMap(
    ColumnImpl.fromValues(availableColumns as Comparable[]),
    columnMap
  );

  return all(filter(sortedColumns.values(), Boolean)) as Promise<Column[]>;
};
