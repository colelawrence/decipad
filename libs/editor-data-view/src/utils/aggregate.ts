import { debounceTime } from 'rxjs/operators';
import { EMPTY } from 'rxjs';
import { ColumnLike, Comparable } from '@decipad/column';
import {
  RemoteComputer,
  Result,
  SerializedType,
} from '@decipad/remote-computer';
import { PreviousColumns } from '../types';
import { textify } from '@decipad/parse';
import { maybeAggregate } from './maybeAggregate';

export interface AggregateProps {
  computer: RemoteComputer;
  tableName?: string;
  aggregationType?: string;
  column?: {
    type: SerializedType;
    value: ColumnLike<Comparable>;
    name: string;
  };
  previousColumns: PreviousColumns;
  roundings: Array<string | undefined>;
}

export interface AggregationResult {
  result: Result.Result;
  expression?: string | Error;
}

const DEBOUNCE_RESULT_MS = 100;

export const aggregate = ({
  computer,
  tableName,
  aggregationType,
  column,
  previousColumns,
  roundings,
}: AggregateProps): Promise<AggregationResult> => {
  return new Promise((resolve) => {
    const expressionFilter =
      (column &&
        previousColumns.reduce((previous, current, index) => {
          const rounding = roundings[index];
          const filterSubject = previous
            ? `${previous}.${current.name}`
            : `${tableName}.${current.name}`;
          const roundedFilterSubject = rounding
            ? `round(${filterSubject}, ${rounding})`
            : filterSubject;
          const escapedValue = textify({
            type: current.type,
            value: current.value as Result.Result['value'],
          });
          return previous === ''
            ? `filter(${tableName}, ${roundedFilterSubject} == ${escapedValue})`
            : `filter(${previous}, ${roundedFilterSubject} == ${escapedValue})`;
        }, '')) ||
      tableName;

    const expression =
      column &&
      expressionFilter &&
      maybeAggregate(
        `${expressionFilter}.${column.name}`,
        column.type,
        aggregationType,
        {
          sum: `sum(${tableName}.${column.name})`,
        }
      );

    const sub = (
      (typeof expression === 'string' &&
        expression &&
        computer
          .expressionResultFromText$(expression)
          .pipe(debounceTime(DEBOUNCE_RESULT_MS))) ||
      EMPTY
    ).subscribe((result) => {
      sub.unsubscribe();
      resolve({ result, expression });
    });
  });
};
