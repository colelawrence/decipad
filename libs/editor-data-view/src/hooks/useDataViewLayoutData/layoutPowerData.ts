import { type Result } from '@decipad/remote-computer';
import { Value } from '@decipad/language-types';
import { type DataViewFilter } from '@decipad/editor-types';
import type { AggregationKind, DataGroup } from '../../types';
import { generateGroups } from './generateGroups';
import { generateSmartRow } from './generateSmartRow';
import { identity } from '@decipad/utils';

interface LayoutPowerDataProps {
  tableName: string;
  tree: Result.Result<'tree'>;
  aggregationTypes: Array<AggregationKind | undefined>;
  roundings: Array<string | undefined>;
  filters: Array<DataViewFilter | undefined>;
  expandedGroups: string[] | undefined;
  includeTotal?: boolean;
  preventExpansion: boolean;
}

export const layoutPowerData = async ({
  tableName,
  tree,
  aggregationTypes,
  roundings,
  filters,
  expandedGroups,
  includeTotal = true,
  preventExpansion = false,
}: LayoutPowerDataProps): Promise<DataGroup[]> => {
  if (!(tree.value instanceof Value.Tree)) {
    console.error('Expected tree value', tree);
    throw new Error('Expected tree value');
  }
  const rootGroups = await generateGroups({
    tableName,
    tree,
    expandedGroups,
    preventExpansion,
    valuePath: [],
    previousColumns: [],
    previousColumnTypes: [],
    previousFilters: [],
    aggregations: aggregationTypes,
    roundings,
    filters,
  });

  const totalGroup =
    includeTotal &&
    tree.value.columns.length > 0 &&
    generateSmartRow({
      aggregation: tree.value.columns[0].aggregation,
      tableName,
      columnTypes: tree.type.columnTypes,
      columns: tree.value.columns,
      global: true,
      valuePath: [],
      previousColumns: [],
      previousColumnTypes: [],
      previousFilters: [],
      aggregations: aggregationTypes,
      roundings,
      filters,
      indent: -Infinity,
    });

  const result = [
    ...rootGroups,
    ...(totalGroup && !!aggregationTypes.filter(identity).length
      ? [totalGroup]
      : []),
  ];

  return result;
};
