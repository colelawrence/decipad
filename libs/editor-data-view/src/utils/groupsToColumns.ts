import {
  RemoteComputer,
  Result,
  SerializedType,
  Unknown,
} from '@decipad/remote-computer';
import { AggregationKind, Column, DataGroup } from '../types';
import { aggregate } from './aggregate';

const concatenate = (a: Column[], b: Column[]): Column[] => {
  return b.map((col, i) => ({
    ...col,
    value: (a[i]?.value ?? []).concat(col.value),
  }));
};

export interface GroupsToColumnsOptions {
  computer: RemoteComputer;
  tableName?: string;
  groups: DataGroup[];
  aggregationTypes: (AggregationKind | undefined)[];
  roundings: Array<string | undefined>;
}

// eslint-disable-next-line complexity
export const groupsToColumns = async ({
  computer,
  tableName,
  groups,
  aggregationTypes,
  roundings,
}: GroupsToColumnsOptions): Promise<Column[]> => {
  let type: SerializedType | undefined;
  let values: Result.OneResult[] = [];
  let columnName: string | undefined;
  let childrenColumns: Column[] = [];

  for (const group of groups) {
    if (group.id === 'total') {
      continue;
    }
    if (group.elementType === 'group') {
      // type
      if (!type && group.type) {
        type = group.type;
      }

      // value
      const value: Result.OneResult =
        (group.value as Result.OneResult) ?? Unknown;
      if (Array.isArray(value)) {
        values = values.concat(value);
      } else {
        values.push(value);
      }

      // name
      if (!columnName && group.column) {
        columnName = group.column.name;
      }
    } else {
      // group.elementType === 'smartrow'
      const { name, type: columnType } = group.column;
      if (!columnName) {
        columnName = name;
      }
      if (!type && columnType) {
        type = columnType;
      }
      const aggregationType = aggregationTypes[0];
      if (!aggregationType) {
        values.push(Unknown);
      } else {
        // eslint-disable-next-line no-await-in-loop
        const { result } = await aggregate({
          computer,
          tableName,
          aggregationType,
          column: group.column,
          previousColumns: group.previousColumns,
          roundings,
        });
        if (result.value != null) {
          values.push(result.value);
        }
      }
    }

    // recurse next columns
    if (group.children.length) {
      childrenColumns = concatenate(
        childrenColumns,
        // eslint-disable-next-line no-await-in-loop
        await groupsToColumns({
          computer,
          tableName,
          groups: group.children,
          aggregationTypes: aggregationTypes.slice(1),
          roundings: roundings.slice(1),
        })
      );
    } else if (aggregationTypes.length > 1) {
      const missing: Array<undefined> = Array.from({
        length: aggregationTypes.length - 1,
      });
      childrenColumns = concatenate(
        childrenColumns,
        // eslint-disable-next-line no-await-in-loop
        await groupsToColumns({
          computer,
          tableName,
          groups: [
            {
              elementType: 'group',
              children: [],
              previousColumns: [],
              replicaCount: 1,
              columnIndex: 0,
            },
          ],
          aggregationTypes: missing,
          roundings: missing,
        })
      );
    }

    // extend current column to match length of next column
    const nextColumn = childrenColumns[0];
    if (nextColumn && values.length) {
      while (nextColumn.value.length > values.length) {
        values.push(values[values.length - 1]);
      }
      while (nextColumn.value.length < values.length) {
        nextColumn.value.push(Unknown);
      }
    }
  }
  if (!values.length) {
    return [];
  }
  return [
    {
      type: type ?? { kind: 'pending' },
      value: values.length ? values : [Unknown],
      name: columnName ?? 'unknown',
    },
    ...childrenColumns,
  ];
};
