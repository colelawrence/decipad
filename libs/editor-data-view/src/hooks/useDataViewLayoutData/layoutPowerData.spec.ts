import { expect, describe, it } from 'vitest';
import { nanoid } from 'nanoid';
import { type Result, type AST, Unknown } from '@decipad/language-interfaces';
import type { IdentifiedBlock, Program } from '@decipad/computer-interfaces';
import { Value } from '@decipad/language-types';
import { getDefined } from '@decipad/utils';
import { buildResult, decilang, getComputer } from '@decipad/computer';
import { layoutPowerData } from './layoutPowerData';

const programBlock = (
  statement: AST.Statement,
  blockId = nanoid()
): IdentifiedBlock => {
  return {
    type: 'identified-block',
    id: blockId,
    block: {
      id: blockId,
      type: 'block',
      args: [statement],
    },
  };
};

const toProgram = (code: AST.Statement[]): Program => {
  return code.map((c, index) => programBlock(c, `block-${index}`));
};

describe('layoutPowerData', () => {
  it('works with an empty tree', async () => {
    expect(
      await layoutPowerData({
        tableName: 'TableName',
        tree: buildResult(
          {
            kind: 'tree',
            columnTypes: [],
            columnNames: [],
          },
          Value.Tree.from(Unknown, undefined, [], [])
        ),
        aggregationTypes: [],
        roundings: [],
        filters: [],
        expandedGroups: [],
        preventExpansion: false,
      })
    ).toMatchInlineSnapshot(`[]`);
  });

  it('works with one column one group tree', async () => {
    const computer = getComputer();
    const results = await computer.computeDeltaRequest({
      program: {
        upsert: toProgram([
          decilang`DataTable = {
          Column1 = ["A"]
        }`,
          decilang`Tree = tree(DataTable)`,
        ]),
      },
    });
    const tree = getDefined(
      results?.blockResults['block-1']?.result,
      'no result for block-1'
    ) as Result.Result<'tree'>;
    expect(tree).toMatchInlineSnapshot(`
      {
        "meta": undefined,
        "type": {
          "columnNames": [
            "Column1",
          ],
          "columnTypes": [
            {
              "kind": "string",
            },
          ],
          "kind": "tree",
        },
        "value": Tree {
          "children": [
            Tree {
              "children": [],
              "columns": [],
              "originalCardinality": 1,
              "root": "A",
              "rootAggregation": undefined,
            },
          ],
          "columns": [
            {
              "aggregation": undefined,
              "name": "Column1",
            },
          ],
          "originalCardinality": 1,
          "root": Symbol(unknown),
          "rootAggregation": undefined,
        },
      }
    `);
    expect(
      await layoutPowerData({
        tableName: 'TableName',
        tree,
        aggregationTypes: ['string:count-unique'],
        expandedGroups: [],
        roundings: [],
        filters: [],
        preventExpansion: false,
      })
    ).toMatchInlineSnapshot(`
      [
        {
          "aggregationExpression": "count(unique((filter(TableName, TableName.Column1 == "A")).Column1))",
          "aggregationResult": undefined,
          "children": [],
          "collapsible": false,
          "elementType": "group",
          "id": "/a32fd8b31587b9c6354faa86e12756d2e7aca860346be182aacebb91f0b58026",
          "replicaCount": 1,
          "type": {
            "kind": "string",
          },
          "value": "A",
        },
        {
          "aggregationExpression": "count(unique(TableName.Column1))",
          "children": [],
          "elementType": "smartrow",
          "global": true,
          "type": undefined,
          "value": Symbol(unknown),
        },
      ]
    `);
  });

  it('works with one column one group tree and aggregation', async () => {
    const computer = getComputer();
    const results = await computer.computeDeltaRequest({
      program: {
        upsert: toProgram([
          decilang`DataTable = {
          Column1 = ["A"]
        }`,
          decilang`Empty = {}`,
          decilang`Aggregate(x) = count(unique(x))`,
          decilang`Aggregations = {
          Column1 = Aggregate
        }`,
          decilang`Tree = tree(DataTable, Empty, Empty, Aggregations)`,
        ]),
      },
    });
    const tree = getDefined(
      results?.blockResults['block-4']?.result,
      'no result for block-4'
    ) as Result.Result<'tree'>;
    expect(tree).toMatchInlineSnapshot(`
      {
        "meta": undefined,
        "type": {
          "columnNames": [
            "Column1",
          ],
          "columnTypes": [
            {
              "kind": "string",
            },
          ],
          "kind": "tree",
        },
        "value": Tree {
          "children": [
            Tree {
              "children": [],
              "columns": [],
              "originalCardinality": 1,
              "root": "A",
              "rootAggregation": undefined,
            },
          ],
          "columns": [
            {
              "aggregation": {
                "meta": undefined,
                "type": {
                  "kind": "number",
                  "unit": null,
                },
                "value": DeciNumber {
                  "d": 1n,
                  "infinite": false,
                  "n": 1n,
                  "s": 1n,
                },
              },
              "name": "Column1",
            },
          ],
          "originalCardinality": 1,
          "root": Symbol(unknown),
          "rootAggregation": undefined,
        },
      }
    `);
    expect(
      await layoutPowerData({
        tableName: 'TableName',
        tree,
        aggregationTypes: ['string:count-unique'],
        roundings: [],
        filters: [],
        expandedGroups: [],
        preventExpansion: false,
      })
    ).toMatchInlineSnapshot(`
      [
        {
          "aggregationExpression": "count(unique((filter(TableName, TableName.Column1 == "A")).Column1))",
          "aggregationResult": undefined,
          "children": [],
          "collapsible": false,
          "elementType": "group",
          "id": "/a32fd8b31587b9c6354faa86e12756d2e7aca860346be182aacebb91f0b58026",
          "replicaCount": 1,
          "type": {
            "kind": "string",
          },
          "value": "A",
        },
        {
          "aggregationExpression": "count(unique(TableName.Column1))",
          "children": [],
          "elementType": "smartrow",
          "global": true,
          "type": {
            "kind": "number",
            "unit": null,
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 1n,
            "s": 1n,
          },
        },
      ]
    `);
  });

  it('works with two column two groups tree and aggregation', async () => {
    const computer = getComputer();
    const results = await computer.computeDeltaRequest({
      program: {
        upsert: toProgram([
          decilang`DataTable = {
          Column1 = ["A", "B", "A"]
          Column2 = [3, 2, 1]
        }`,
          decilang`Empty = {}`,
          decilang`Aggregate1(x) = count(unique(x))`,
          decilang`Aggregate2(x) = sum(x)`,
          decilang`Aggregations = {
          Column1 = Aggregate1
          Column2 = Aggregate2
        }`,
          decilang`Tree = tree(DataTable, Empty, Empty, Aggregations)`,
        ]),
      },
    });
    const tree = getDefined(
      results?.blockResults['block-5']?.result,
      'no result for block-5'
    ) as Result.Result<'tree'>;
    expect(tree).toMatchSnapshot();
    expect(
      await layoutPowerData({
        tableName: 'TableName',
        tree,
        aggregationTypes: ['string:count-unique', 'number:sum'],
        roundings: [],
        filters: [],
        expandedGroups: [],
        preventExpansion: false,
      })
    ).toMatchInlineSnapshot(`
      [
        {
          "aggregationExpression": "count(unique((filter(TableName, TableName.Column1 == "A")).Column1))",
          "aggregationResult": undefined,
          "children": [
            {
              "aggregationExpression": "sum((filter(TableName, TableName.Column1 == "A")).Column2)",
              "children": [],
              "elementType": "smartrow",
              "global": false,
              "type": {
                "kind": "number",
                "unit": null,
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 4n,
                "s": 1n,
              },
            },
          ],
          "collapsible": true,
          "elementType": "group",
          "id": "/ab2343f77c33a58d1431ce0434a21d3d385f00b3308206522aaef176e539a7cc",
          "replicaCount": 2,
          "type": {
            "kind": "string",
          },
          "value": "A",
        },
        {
          "aggregationExpression": "count(unique((filter(TableName, TableName.Column1 == "B")).Column1))",
          "aggregationResult": undefined,
          "children": [
            {
              "aggregationExpression": "sum((filter(filter(TableName, TableName.Column1 == "B"), (filter(TableName, TableName.Column1 == "B")).Column2 == 2)).Column2)",
              "aggregationResult": undefined,
              "children": [],
              "collapsible": false,
              "elementType": "group",
              "id": "/e6058265454b52c33d07e0394ea71bcebe74c4c386cf92d36494fa1c23fbfed9/b0bc58bdde052273dc781192e21d2539e3dc453f1364463569428c0590b16f32",
              "replicaCount": 1,
              "type": {
                "kind": "number",
                "unit": null,
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 2n,
                "s": 1n,
              },
            },
          ],
          "collapsible": false,
          "elementType": "group",
          "id": "/e6058265454b52c33d07e0394ea71bcebe74c4c386cf92d36494fa1c23fbfed9",
          "replicaCount": 1,
          "type": {
            "kind": "string",
          },
          "value": "B",
        },
        {
          "aggregationExpression": "count(unique(TableName.Column1))",
          "children": [
            {
              "aggregationExpression": "sum(TableName.Column2)",
              "children": [],
              "elementType": "smartrow",
              "global": true,
              "type": {
                "kind": "number",
                "unit": null,
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 6n,
                "s": 1n,
              },
            },
          ],
          "elementType": "smartrow",
          "global": true,
          "type": {
            "kind": "number",
            "unit": null,
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 2n,
            "s": 1n,
          },
        },
      ]
    `);
  });
});
