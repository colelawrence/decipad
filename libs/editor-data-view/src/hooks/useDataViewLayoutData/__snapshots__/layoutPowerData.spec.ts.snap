// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layoutPowerData > works with two column two groups tree and aggregation 1`] = `
{
  "type": {
    "columnNames": [
      "Column1",
      "Column2",
    ],
    "columnTypes": [
      {
        "kind": "string",
      },
      {
        "kind": "number",
        "unit": null,
      },
    ],
    "kind": "tree",
  },
  "value": Tree {
    "children": [
      Tree {
        "children": [
          Tree {
            "children": [],
            "columns": [],
            "originalCardinality": 1,
            "root": DeciNumber {
              "d": 1n,
              "infinite": false,
              "n": 1n,
              "s": 1n,
            },
            "rootAggregation": undefined,
          },
          Tree {
            "children": [],
            "columns": [],
            "originalCardinality": 1,
            "root": DeciNumber {
              "d": 1n,
              "infinite": false,
              "n": 3n,
              "s": 1n,
            },
            "rootAggregation": undefined,
          },
        ],
        "columns": [
          {
            "aggregation": {
              "type": {
                "kind": "number",
                "unit": null,
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 4n,
                "s": 1n,
              },
            },
            "name": "Column2",
          },
        ],
        "originalCardinality": 2,
        "root": "A",
        "rootAggregation": {
          "type": {
            "kind": "number",
            "unit": null,
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 2n,
            "s": 1n,
          },
        },
      },
      Tree {
        "children": [
          Tree {
            "children": [],
            "columns": [],
            "originalCardinality": 1,
            "root": DeciNumber {
              "d": 1n,
              "infinite": false,
              "n": 2n,
              "s": 1n,
            },
            "rootAggregation": undefined,
          },
        ],
        "columns": [
          {
            "aggregation": {
              "type": {
                "kind": "number",
                "unit": null,
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 2n,
                "s": 1n,
              },
            },
            "name": "Column2",
          },
        ],
        "originalCardinality": 1,
        "root": "B",
        "rootAggregation": {
          "type": {
            "kind": "number",
            "unit": null,
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 1n,
            "s": 1n,
          },
        },
      },
    ],
    "columns": [
      {
        "aggregation": {
          "type": {
            "kind": "number",
            "unit": null,
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 2n,
            "s": 1n,
          },
        },
        "name": "Column1",
      },
      {
        "aggregation": {
          "type": {
            "kind": "number",
            "unit": null,
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 6n,
            "s": 1n,
          },
        },
        "name": "Column2",
      },
    ],
    "originalCardinality": 1,
    "root": Symbol(unknown),
    "rootAggregation": undefined,
  },
}
`;
